service/tbc/tbc.go

```diff
@@ -1089,43 +1258,36 @@ func (s *Server) handleHeaders(ctx context.Context, p *peer, msg *wire.MsgHeader
 	// There really is no good way of determining if we can escape the
 	// expensive calls so we just eat it.
 	var pbhHash *chainhash.Hash
-	headers := make([][80]byte, len(msg.Headers))
 	for k := range msg.Headers {
 		if pbhHash != nil && pbhHash.IsEqual(&msg.Headers[k].PrevBlock) {
-			log.Errorf("cannot connect %v index %v",
+			return fmt.Errorf("cannot connect %v index %v",
 				msg.Headers[k].PrevBlock, k)
-			p.close() // get rid of this misbehaving peer
-			return
 		}
-
-		copy(headers[k][0:80], h2b(msg.Headers[k])) // XXX don't double copy
 		pbhHash = &msg.Headers[k].PrevBlock
 	}
-
-	it, cbh, lbh, err := s.db.BlockHeadersInsert(ctx, headers)
+	it, cbh, lbh, n, err := s.db.BlockHeadersInsert(ctx, msg)
 	if err != nil {
 		// This ends the race between peers during IBD.
 		if errors.Is(database.ErrDuplicate, err) {
-			// XXX for now don't do paralle blockheader downloads.
+			// XXX for now don't do parallel blockheader downloads.
 			// Seems to really slow the process down.
 			//
 			// We already have these headers. Ask for best headers
 			// despite racing with other peers. We do that to
 			// prevent stalling the download.
 			//bhb, err := s.db.BlockHeaderBest(ctx)
 			//if err != nil {
-			//	log.Errorf("block headers best %v: %v", p, err)
+			//	log.Errorf("block header best %v: %v", p, err)
 			//	return
 			//}
 			//if err = s.getHeaders(ctx, p, bhb.Header); err != nil {
 			//	log.Errorf("get headers %v: %v", p, err)
 			//	return
 			//}
-			return
+			return nil
 		}
 		// Real error, abort header fetch
-		log.Errorf("block headers insert: %v", err)
-		return
+		return fmt.Errorf("block headers insert: %v", err)
```

Comment:
I think the `%w` verb can be used here to wrap the error (and in all of the `fmt.Errorf`s below):
```suggestion
		return fmt.Errorf("block headers insert: %w", err)
```